### PR TITLE
bugfix/missing quotes templates

### LINKED ISSUE #129: Count path segments are missing after translation
[OData supports counts segements](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_AddressingtheCountofaCollection) by conventions. e.g. GET /users/$count

All these segments are missing from the resulting OpenAPI description.

### LINKED ISSUE #140: URL templates missing quotes around enum types
According to the spec, it seems that we need quotes around the enum type name but not the namespace:
http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_PrimitiveLiterals
> `ColorEnumValue eq Sales.Pattern'Yellow',`
> `ColorEnumValue eq 'Yellow',`

These quotes are missing in the translated OpenAPI doc.

see https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/1156 for more details.

### PR DESCRIPTION
**Pull request by baywet**
- fixes #140 a bug where quotes would be missing in url templates

### CODE REVIEW

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/TripService.OpenApi.yaml [hunk 1]
```diff
@@ -390,7 +390,7 @@ paths:
         default:
           $ref: '#/components/responses/error'
       x-ms-docs-operation-type: operation
-  '/GetNearestAirport(lat={lat},lon={lon})':
+  '/GetNearestAirport(lat=''{lat}'',lon=''{lon}'')':
```
Comment (zengin) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/TripService.OpenApi.yaml:393]: same comment but for double.

**Review by zengin — CHANGES_REQUESTED** [review]

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataOperationSegmentTests.cs [hunk 2]
```diff
@@ -74,10 +74,10 @@ public void GetPathItemNameReturnsCorrectActionLiteral(bool unqualifiedCall, boo
         }
 
         [Theory]
-        [InlineData(true, true, "MyFunction(param={param})")]
-        [InlineData(true, false, "MyFunction(entity=@entity,param={param})")]
-        [InlineData(false, true, "NS.MyFunction(param={param})")]
-        [InlineData(false, false, "NS.MyFunction(entity=@entity,param={param})")]
+        [InlineData(true, true, "MyFunction(param='{param}')")]
```
[the remark this answers is not in the document]
Comment (zengin) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataOperationSegmentTests.cs:77]: http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_PrimitiveLiterals

[the remark this answers is not in the document]
Comment (xuzhg) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataOperationSegmentTests.cs:77]: single quote is used for string primitive type value in the request path. 
for the template, I don't understand why should we have a single quote around.
It's just a template and anything within the {} is the parameter name used to retrieve the value from request.

**Review by xuzhg — CHANGES_REQUESTED** [review]

**Comment by xuzhg** [comment]
`MyFunction(param={param})` can match

- ~ MyFunction(param='abc')  // if the type of param is string
- ~ MyFunction(param=true) // if the type of param is boolean
- ~ MyFunction(param=8) // if the type of param is int32
- ...
- 

MyFunction(param='{param}') can matches ?

- ~ MyFunction(param=''abc'')  // if the type of param is string
- ~ MyFunction(param='true') // if the type of param is boolean
- ~ MyFunction(param='8') // if the type of param is int32

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataOperationSegmentTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (baywet) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataOperationSegmentTests.cs]: thank you both for catching my replacement mistake! I reviewed my copy and just pushed a version which I believe will be more accurate, let me know what you think.

**Review by zengin — APPROVED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+{
+	/// <summary>
+    /// Extension methods for <see cref="IEdmType"/>
+    /// </summary>
+    public static class EdmTypeExtensions
+	{
+		/// <summary>
+		/// Determines wether a path parameter should be wrapped in quotes based on the type of the parameter.
+		/// </summary>
+		/// <param name="edmType">The type of the parameter.</param>
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: format

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataKeySegment.cs [hunk 4]
```diff
@@ -77,13 +77,7 @@ public override string GetPathItemName(OpenApiConvertSettings settings, HashSet<
                 }
                 else
                 {
-                    IList<string> keyStrings = new List<string>();
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataKeySegment.cs:80]: Again, I have a different opinion for the single quote.

Single quote is part of the string, enum value, it goes with value. But not for the template.

When Graph SDK generate the URL (the real URL), should generate the single quote for string type. 
For example:

'8' for string type, but 8 for int type. 

So, the changes in this PR looks "invalid" to me. I'd like @mikepizzo's input.

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataKeySegment.cs:80]: I disagree. This quotes behavior comes from odata conventions. The open api spec does not mandate any quote for url parameters, regardless of the types.
The point of this library is to encapsulate the odata conventions away.
This is why I believe the quotes should be added during the conversion process.

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs (hunk 3 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: Sorry, what's the issue with the format here?

Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: alignment, thanks.

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: are you referring to the use of tabs vs spaces? still trying to understand what you think needs to be changed here.

Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: Yes, replace tab using 4 spaces.

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/EdmTypeExtensions.cs:18]: done!

**Comment by darrelmiller** [comment]
The OpenAPI that we need to generate should look like this,

```yaml
openapi: 3.0.0
info:
  title: example of string path parameter that must be quoted
  version: 1.0.0
paths:
  get:
    /MyFunction(param='{funcParam}'):
      parameters:
        - name: funcParam
          in: path
          required: true
          schema:
             type: string
      responses: 
        200:
          description: OK
```

JSON Schema has no notion of a quoted string, so it would be difficult to describe a parameter value that needs to be surrounded by quotes before inserting into the URL.  The fact that quotes, single vs double are required are an OData requirement and therefore should be encoded in URL, just like all other OData conventions.

You can see in the simple string expansion examples in the URI Template specification that quotes are not inserted for string values.

```code
     Example Template     Expansion

       {var}              value
       {hello}            Hello%20World%21
       {half}             50%25
       O{empty}X          OX
       O{undef}X          OX
       {x,y}              1024,768
       {x,hello,y}        1024,Hello%20World%21,768
       ?{x,empty}         ?1024,
       ?{x,undef}         ?1024
       ?{undef,y}         ?768
       {var:3}            val
       {var:30}           value
       {list}             red,green,blue
       {list*}            red,green,blue
       {keys}             semi,%3B,dot,.,comma,%2C
       {keys*}            semi=%3B,dot=.,comma=%2C
```
https://datatracker.ietf.org/doc/html/rfc6570#section-3.2.2

**Comment by baywet** [comment]
collecting notes from the meeting we've just had:

- clients rely on this conversion library to encapsulate away OData conventions and therefore have no knowledge of encoding requirements (quotes)
- the information (quotes) cannot be carried through JSON schema/OpenAPI structures
- this should work in 99.999% of the times, except maybe for Edm.Primitive types which are not allowed in url templates
- the tradeoff of this solution is making the template a little more static in nature
- an alternative could be considered to use the format property, and use an open formatter (odatavalue, odatastring...) but this brings other concerns like collisions with numbers which already use the format property.

Overall the group agreed to proceed with this change.

@xuzhg for final review and merge.

**Comment by xuzhg** [comment]
:shipit:

**Pull request merged**